extensions/debug-server-ready/src/extension.ts

```diff
@@ -183,14 +191,78 @@ class ServerReadyDetector extends vscode.Disposable {
 		}
 	}
 
-	private debugWithBrowser(type: string, session: vscode.DebugSession, uri: string) {
+	private async debugWithBrowser(type: string, session: vscode.DebugSession, uri: string) {
+		const args = session.configuration.serverReadyAction as ServerReadyAction;
+		if (!args.killOnServerStop) {
+			await this.startBrowserDebugSession(type, session, uri);
+			return;
+		}
+
+		const trackerId = randomUUID();
+		const newSessionPromise = new Promise<vscode.DebugSession>(resolve => {
+			const listener = vscode.debug.onDidStartDebugSession(newSession => {
+				if (newSession.configuration.trackerId === trackerId) {
+					listener.dispose();
+					resolve(newSession);
+				}
+			});
+			this.lateDisposables.push(listener); // In case `trackerId` of interest was never caught anyhow.
+		});
```

Comment:
Applicable to here and below:

- The listener will be leaked if `startDebugging` returns false
- Even after the listener resolves, `lateDisposables` will keep growing indefinitely.

Maybe `lateDisposables` should be a `Set` that you can delete disposables from.

Reply:
You're right. Will push a fix soon.

Reply:
Made the changes as you suggested. the `lateDisposables` is a `Set` now. Also used `vscode.CancellationToken` for signalling whether to stop listening for events anymore.